### ISSUE TITLE
How to truncate the preview results on `files_resume`, `live_grep_resume`, etc.?

### ISSUE THREAD

**Issue by nyngwang** [ibhagwan/fzf-lua#261]
TL;DR: The status-quo is great. But I still expect that it can be faster.

While the `multiprocess = true` option is great, it seems that I still experience some delay when scrolling down/up the result list with `<C-J/K>` on `files_resume`, `live_grep_resume`, etc. I guess that this kind of delay is inevitable since the previewer needs to prepare the entire file to be displayed. So I want to ask how to add options so that it only shows 10 lines around the found word in the preview area when using `live_grep_resume` and that it only shows the first 30 lines of code on that when using `files_resume`. I'm not sure whether this is related to `fzf`, `fd`, `rg` or `cat` and where I should add the option(s) to achieve this result (if possible).

**Comment by ibhagwan** [comment]
> While the `multiprocess = true` option is great, it seems that I still experience some delay when scrolling down/up the result list with `<C-J/K>` on `files_resume`, `live_grep_resume`, etc. I guess that this kind of delay is inevitable since the previewer needs to prepare the entire file to be displayed. 

There is a 100ms delay by design, it's the same as running fzf in the shell with a `bat` preview in order not to spam read previews while scrolling.

In fzf-lua you can control the preview delay or even set it to `0` in the setup under `winopts.preview.delay`, you can try and see how it feels with this:
```lua
:lua require'fzf-lua'.files({ winopts = { preview = { delay = 0 } } })
```

>So I want to ask how to add options so that it only shows 10 lines around the found word in the preview area when using `live_grep_resume` and that it only shows the first 30 lines of code on that when using `files_resume`. I'm not sure whether this is related to `fzf`, `fd`, `rg` or `cat` and where I should add the option(s) to achieve this result (if possible).

You can't do that with the builtin previewer but you can craft your own fzf native previewer if you wanted that runs an `rg` command with the `--after-context` and `--before-context` options but you'll lose syntax highlighting, you may be able to do that with `bat` as well but I haven't checked.

**Comment by nyngwang** [comment]
Thanks for your responsiveness first :) For the first part, I'm happy to try the option! Regarding the second part, I used to suspect that `bat` would consume more system resources (but I didn't do any experiment yet) so I resort to the builtin one. And the hand-craft-myself way is impossible for me currently.

**Comment by nyngwang** [comment]
> Thanks for your quick reply first :) For the first part, I'm happy to try the option! **(Added on edit) But does it has any side-effect if I set the number too small?** [...]

After a try, I think tuning the `delay` is the best solution for now. Now I would like to know the question I quoted.

**Comment by nyngwang** [comment]
@ibhagwan I just come up with an idea: Is it worth implementing this heuristic?

> If a user doesn't keep the cursor long enough on an entry of a search result. Don't run the command to display the result.

There might be many problems to consider though since I know nothing about the implementation details.

**Comment by ibhagwan** [comment]
> But does it has any side-effect if I set the number too small?

If the number is too small it just means the preview buffer will be read more frequently, this may or may not have some overhead depending on the way you use the picker as you might be reading more files than you require, no big deal either way, even 20ms delay should be enough to provide some protection against spam reads when pressing down arrow non stop.

>If a user doesn't keep the cursor long enough(say 0.5 sec.) on an entry of a search result, then don't run the (preview) command for that result, since the user will not care about previewing those files that will just get "pass-by". No need to spend time on this part.

Unfortunately this information simply doesn’t exist in this context, the only “event” I receive from fzf is when the cursor appears on a file, I have no idea which key was pressed (down arrow or ctrl-j or search typing) and no way of knowing if the key was pressed or not.

I’m not even sure what you’re trying to solve here?

**Comment by nyngwang** [comment]
> I’m not even sure what you’re trying to solve here?

In my current use case, after typing some keywords as the input for `live_grep_resume` or `files_resume`, I will not add more keywords anymore, and my next step will be to find the one I'm interested in the result list. I will not care about previewing those files I'm not interested, but the current state is that I experience many "blinks", i.e. files I'm not interested but displayed when my cursor just want to pass-by them.

**Comment by ibhagwan** [comment]
Once you narrowed down your results why don’t you then close the preview with F4?

**Comment by nyngwang** [comment]
That might be a solution since I haven't tried it. I will share my experience using this method.

**Comment by nyngwang** [comment]
@ibhagwan Oh, I'm so stupid. I simply set the delay to `500` and this is exactly the experience we just discuss about!

**Comment by ibhagwan** [comment]
> > In my current use case, after typing some keywords as the input for live_grep_resume or files_resume, I will not add more keywords, but my next step will be to find the one(s) I'm interested in the result list. I will not care about previewing those files I'm not interested but I have to "pass-by". To be more clear, the current state is that I experience many "blinks"(i.e. files I'm not interested in getting previewed), when I just want my cursor to "pass-by" these files to take a look at those files that I think/guess would be what I need.
> 
> @ibhagwan Oh, sorry I'm so stupid. I simply set the delay to `500` and this is exactly the experience I want to achieve! I think this question can be closed! It's just hilarious that my initial direction on solving this problem was trying to set a smaller delay.

Great!

**Comment by nyngwang** [comment]
Now the scrolling becomes very smooth. But if I press and hold the `<C-J>` the computer will get very hot. So I guess that while those files that get pass-by are not previewed some resources are still spent on them. Is this true? If so, how to avoid it?

**Comment by ibhagwan** [comment]
> Now the scrolling becomes very smooth. But if I press and hold the `<C-J>` the computer will get very hot (55 celsius all the way up to 80 celsius). So I guess that while those files that get pass-by are not previewed, some resources are still spent on them. Is this true? If so, how to avoid it?

Idk why that happens, no resources should be spent on these files at all, can you try running the same command in the shell and pipe it to `fzf --ansi` and see if you get the same issue? if you do that might be an `fzf` issue which isn't related to fzf-lua, if that's the case you can open an issue [upstream](https://github.com/junegunn/fzf).

You can also try with the default previewer set to `bat`, that way all the functionality is offloaded to fzf (fzf-lua and neovim do nothing after the command is complete) but note that if you do that the delay will be 100ms which is set internall by fzf which I can't control.

**Comment by ibhagwan** [comment]
Also, you can try [`skim`](https://github.com/lotabout/skim) instead of `fzf`, maybe it doesn't have this issue?

Once installed all you need to do is add `fzf_bin = 'sk'` to the `setup` options.

**Comment by nyngwang** [comment]
> Also, you can try skim instead of fzf, maybe it doesn't have this issue?

Do I have to change/remove my own `fzf_opts`/`rg_opts` when trying it?

**Comment by nyngwang** [comment]
Not sure if this is related but I just removed the `fzf-nvim` from the dependencies after checking the README.md again:

```
requires = {
  -- WARNING: nvim-fzf is not fzf itself, install it using brew!
  -- 'vijaymarupudi/nvim-fzf',
  'kyazdani42/nvim-web-devicons'
},
```

**Comment by nyngwang** [comment]
I just tried `sk`, the peak becomes 92 celsius...

**Comment by ibhagwan** [comment]
> I just tried `sk`, the peak becomes 92 celsius...

I'm not sure why this happens, for me it's the opposite, once files stop coming and the loading indicator stops I can scroll as fast as I want and there's no more CPU activity, the work was already done and it's just scrolling in memory.

You need to debug this in the shell, run the same `rg` or `fd` command and pipe it into `fzf` and see how it behaves in the shell, however it behaves in the shell you can replicate with fzf-lua, it's just a terminal command at the end.

If you're still having the same issues in the shell your best bet is to tackle this with the upstream projects.

**Comment by ibhagwan** [comment]
>Not sure if this is related but I just removed the fzf-nvim from the dependencies after checking the README.md again:

This has absolutely nothing to do with it.

Just tried spam scrolling again, holding keys down, anything I can throw at it, once the files stop arriving my CPU barely goes over 10%.

**Comment by nyngwang** [comment]
> Just tried spam scrolling again, holding keys down, anything I can throw at it, once the files stop arriving my CPU barely goes over 10%.

If it's about CPU usage, then the process called `nvim` is also about 10%. But the problem is the CPU temperature goes to about 80 Celsius.

**Comment by ibhagwan** [comment]
You might have a different issue then, maybe terminal rendering, but seems like something else in play here.

**Comment by nyngwang** [comment]
I just noticed that there is a process called `com.broadcom.mes.systemextension` with CPU usage growing from 5% to 80%. I think this might be the source of the problem.

**Comment by nyngwang** [comment]
I ran some `csrutil diable` and disable that process (which is anti-virus related service). Without any process with high CPU usage, the high temperature problem still exists! I gave up! Maybe NeoVim will become the next bloated VSCode in the future...

**Comment by ibhagwan** [comment]
I honestly don’t think this is neovim, I would try different terminals, alacritty, kitty, iTerm, etc.
